### DISCUSSION TITLE
PyBoard 1.1 Strange UART Behaviour

### DISCUSSION THREAD

**Discussion by Davis-Garrad** [micropython/micropython#10565] [Hardware & Peripherals] [answered]
Hello,

I've written some code to use the UART connection between my PyBoard (v1.1) and a Raspberry Pi (model 3, B) to send commands, data, etc. So far, it works 100% of the time with no issues when I have the PyBoard powered via USB from the RPi. The weird thing is, however, if the PyBoard is powered in any other manner (eg. plugging the USB cord into another computer or just using a 12V battery, no USB), it would seem that the output from the PyBoard is sent mostly correctly (with occasional distortions that were not present in the RPi-PyBoard connected setup) and any output from the RPi UART gets sent right back to the RPi.

Once again, I'll emphasize that, aside from the power setup, there are no changes to wiring or code between these runs, and it works perfectly fine when the RPi is powering the PyBoard. I'm very confused, and would love some feedback about what might be causing either of these issues (power source change causing PyBoard to send wrong characters or the "reflection" of signals back to the RPi) 

Thanks in advance!

**Comment by robert-hh** [comment, marked as answer]
Just to ask for the obvious: did you connect GND between the RPi and PyBoard in addition to TX and RX?

**Reply by Davis-Garrad** [reply]
I haven't - I wasn't aware that this was required. They each have their own GNDs, but I'll try connecting them and see what that gets me.

**Reply by Davis-Garrad** [reply]
Well, that worked. You are a scholar and a saint. I can't really find any information on why this changes anything, could you give some explanation as to how you knew this would work?

**Reply by jimmo** [reply]
Voltages (i.e. the signals on the TX and RX pins) only make sense in comparison to another voltage. GND is that reference.

To make a hopefully-helpful-but-not-quite-accurate analogy, imagine I'm trying to push a light switch on and off (i.e. send a signal), but that light switch is floating on a helium balloon. When I push it it will just move away rather that switching. If I anchor that light switch to the wall and my feet are firmly on the floor, then i can now successfully push the button.

**Reply by jimmo** [reply]
A slightly more accurate picture is that you need (very small amounts of) current to flow to actually make the other side receive your message. Current needs to travel in a loop, and ground is the return path for that loop.

**Reply by Davis-Garrad** [reply]
Yeah, thinking about it like a loop makes much more sense - I always seem to neglect that current needs somewhere to go when I've got the path "ending" at the port. Thank you for the responses!

**Comment by robert-hh** [comment]
@jimmo I like the example of the switch at a helium balloon. 
@Davis-Garrad The reason it worked when the device where connected by USB: The usb cables provided the GND connection, and at the speed and signal levels of UART the length of that connection is not very critical. It worked later without GND to some extend, because the other signal wire provided a weak reference potential.

**Reply by Davis-Garrad** [reply]
Thank you! The USB cable providing the common ground makes a lot of sense when considering what @jimmo was saying.

**Comment by peterhinch** [comment]
@Davis-Garrad Ensuring a reliable common ground between all subsystems is crucial in any electronics project. If a subsystem is "floating" it can actually cause physical damage when connected to a grounded subsystem. Explaining the mechanism would require going beyond the rather novel helium balloon analogy, but I'll do so if you wish...

**Reply by Davis-Garrad** [reply]
I'm always happy to hear more explanations for things - especially when they might involve breaking things... However, each system (the PyBoard and RPi) were both connected to their own (separate) power sources' grounds. Intuitively, I can sort of see how if one system is grounded and another isn't, that that would cause problems, but what about two systems that are grounded separately?

**Reply by peterhinch** [reply]
You must have a direct link. There are hazards with separate grounds. Not all systems actually have a ground (in the UK at least). It is common for a PSU (power supply unit) to be double insulated with no actual ground connection. Even if the separate PSU's are grounded, the ground connection is long via the building electrical wiring, posing a risk of interference pickup. Further, if one unit is unplugged it is floating, presenting a static electricity hazard to the other unit.
